### DISCUSSION TITLE
Spy's implementation is not being mocked, for a function in the same module

### DISCUSSION THREAD

**Discussion by garyking** [vitest-dev/vitest#5108] [Q&A]
I have a file called `c.ts`:

```ts
export function printItem(item: string): string {
  if (item === '') {
    throw new Error(`The item cannot be empty.`);
  }

  return 'item';
}

export function printEmptyItem(): void {
  printItem('');
}
```

With a test called `c.test.ts`:

```ts
import { describe, expect, test, vi } from 'vitest';

import * as c from './c.js';

describe('c', () => {
  test('print empty item', () => {
    const spyPrintItem = vi
      .spyOn(c, 'printItem')
      .mockImplementation(() => 'mocked');

    c.printEmptyItem();

    expect(spyPrintItem).toHaveBeenCalledWith('');
  });
});
```

This should mock the `printItem` before `printEmptyItem` is called. However, when I run the test, the error is thrown (`Error: The item cannot be empty.`).

How can I get this to work as intended, so that the `printItem` function is mocked before `printEmptyItem` is called? 

Thanks in advance!

**Comment by sheremet-va** [comment]
It is not possible: https://vitest.dev/guide/mocking.html#mocking-pitfalls

**Reply by garyking** [reply]
Okay thanks. The example code there seems to be wrong though. `return injectedFoo !== foo` returns `true` for me with the exact same code. If I remove `vi.spyOn(mod, 'foo')` then it returns `false`. This seems to indicate that the function is indeed being mocked.

**Reply by sheremet-va** [reply]
It’s a typo in docs, the point is that they are different. They should always be the same if you want to mock it which is impossible.

**Reply by garyking** [reply]
Okay thanks. I moved some code out to other files and the problem was fixed, so it indeed was an issue when both functions were in the same file.
